### PR TITLE
Fix host decoding with punycodes in URLs

### PR DESCRIPTION
**Pull request by lfoppiano**
This PR addresses the issue emerged during the April Crawl where Unicode host names, e.g. `🧠 .s.country` (IDN2008) failed to decode when creating the index table.

### CODE REVIEW

**Comment by lfoppiano** [comment]
@sebastian-nagel after a second though we could make those methods (that I duplicated) public and use them directly from the crawler-commons.

File: src/main/java/org/commoncrawl/net/HostName.java [hunk 1]
```diff
@@ -132,20 +160,30 @@ private void setHostName(String name) {
...
 				try {
-					hostName = URLDecoder.decode(hostName, StandardCharsets.UTF_8.toString());
-				} catch (IllegalArgumentException | UnsupportedEncodingException e) {
-					LOG.error("Failed to decode {}: {}", hostName, e, e.getMessage());
-					hostName = null;
-					return;
+					hostName = URLDecoder.decode(hostName, StandardCharsets.UTF_8);
+				} catch (IllegalArgumentException e) {
+					try {
+						hostName = normalizeName(hostName);
```
Comment (sebastian-nagel) [inline src/main/java/org/commoncrawl/net/HostName.java]: This could be just
```java
IDN.toASCII(hostName, ALLOW_UNASSIGNED);
```
which converts the entire host name to ASCII.

File: src/main/java/org/commoncrawl/net/HostName.java [hunk 2]
```diff
@@ -132,20 +160,30 @@ private void setHostName(String name) {
 			type = Type.hostname;
 			if (hostName.indexOf('%') > -1) {
 				try {
-					hostName = URLDecoder.decode(hostName, StandardCharsets.UTF_8.toString());
-				} catch (IllegalArgumentException | UnsupportedEncodingException e) {
-					LOG.error("Failed to decode {}: {}", hostName, e, e.getMessage());
-					hostName = null;
-					return;
+					hostName = URLDecoder.decode(hostName, StandardCharsets.UTF_8);
```
Comment (sebastian-nagel) [inline src/main/java/org/commoncrawl/net/HostName.java:137]: Good catch to replace the `StandardCharsets.UTF_8.toString()` which is a relict from Java 8.

File: src/main/java/org/commoncrawl/net/HostName.java (hunk 1 above)
Comment (sebastian-nagel) [reply src/main/java/org/commoncrawl/net/HostName.java]: The other point: IDN normalization is not required here, only below if the host name - the original one or the successfully decoded one - contains non-ASCII characters.

File: src/main/java/org/commoncrawl/net/HostName.java [hunk 3]
```diff
@@ -132,20 +160,30 @@ private void setHostName(String name) {
...
 			}
 			if (!CharMatcher.ascii().matchesAllOf(hostName)) {
 				try {
 					hostName = IDN.toASCII(hostName);
 				} catch (IllegalArgumentException | IndexOutOfBoundsException e) {
-					LOG.error("Failed to convert Unicode host name to ASCII {}: {}", hostName, e, e.getMessage());
-					hostName = null;
-					return;
+					try {
+						hostName = normalizeName(hostName);
```
Comment (sebastian-nagel) [inline src/main/java/org/commoncrawl/net/HostName.java]: See comment above.

File: src/test/java/org/commoncrawl/net/HostNameTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+package org.commoncrawl.net;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class HostNameTest {
+	
+	@Test
+	void isAscii_shouldReturnTrue() {
```
Comment (sebastian-nagel) [inline src/test/java/org/commoncrawl/net/HostNameTest.java]: Well, snake case is against the [Java naming conventions](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html).
Maybe could also bundle the single-line unit tests into a single test method.

Anyway: the `isAscii` method is not required and should be replaced by `CharMatcher.ascii().matchesAllOf(str)`. Could just drop them.

File: src/test/java/org/commoncrawl/net/HostNameTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		assertEquals("www.example.com", HostName.normalizeName("WWW.Example.COM"));
+	}
+
+	@Test
+	void normalizeName_mixedAsciiAndUnicode() {
+		assertEquals("www.xn--qv9h.com", HostName.normalizeName("www.🧠.com"));
+	}
+
+	@Test
+	void setHostName_brainEmojiPercentEncoded_isPunycoded() {
```
Comment (sebastian-nagel) [inline src/test/java/org/commoncrawl/net/HostNameTest.java]: See comment above about snake casing.

File: src/main/java/org/commoncrawl/net/HostName.java [hunk 6]
```diff
@@ -114,6 +116,32 @@ public HostName(URI uri) {
 		}
 	}
 
+	static String asciiConvert(String str) throws IllegalArgumentException {
```
Comment (sebastian-nagel) [inline src/main/java/org/commoncrawl/net/HostName.java]: I know that's the code analogous to crawler-commons EffectiveTldFinder where a similar issue has been discovered (crawler-commons/crawler-commons#551). EffectiveTldFinder creates a search tree of public suffixes and splits the suffixes into components, normalizing the components individually.

But here it's simpler to pass the entire host name to `IDN.toASCII` together with the `ALLOW_UNASSIGNED` flag.

**Review by sebastian-nagel — CHANGES_REQUESTED** [review]
Thanks @lfoppiano, the code could be simplified a lot, see the inline comments.

File: src/main/java/org/commoncrawl/net/HostName.java (hunk 1 above)
Comment (lfoppiano) [reply src/main/java/org/commoncrawl/net/HostName.java]: Remove all this part, reduced the battery of unit tests.

**Comment by lfoppiano** [comment]
@sebastian-nagel I've revised the code according to your comments. I also added a test that will generate the hostname to be empty: `normalizeNameNoAsciiInHostname` in `HostnameTest`. Please double check that this is the expected behaviour.

File: src/main/java/org/commoncrawl/net/HostName.java [hunk 7]
```diff
@@ -132,20 +133,23 @@ private void setHostName(String name) {
...
 				}
 			}
 			if (!CharMatcher.ascii().matchesAllOf(hostName)) {
 				try {
 					hostName = IDN.toASCII(hostName);
-				} catch (IllegalArgumentException | IndexOutOfBoundsException e) {
-					LOG.error("Failed to convert Unicode host name to ASCII {}: {}", hostName, e, e.getMessage());
-					hostName = null;
-					return;
+				} catch (IllegalArgumentException e) {
```
Comment (sebastian-nagel) [inline src/main/java/org/commoncrawl/net/HostName.java]: Catching IndexOutOfBoundsException is not a bad idea, see https://bugs.openjdk.org/browse/JDK-6806873 - or is there a specific reason it was removed?

Comment (lfoppiano) [reply src/main/java/org/commoncrawl/net/HostName.java]: Sorry, I missed that.

File: src/main/java/org/commoncrawl/net/HostName.java [hunk 8]
```diff
@@ -141,10 +142,10 @@ private void setHostName(String name) {
 			if (!CharMatcher.ascii().matchesAllOf(hostName)) {
 				try {
 					hostName = IDN.toASCII(hostName);
```
Comment (sebastian-nagel) [inline src/main/java/org/commoncrawl/net/HostName.java]: Sorry, I've missed this in the first review: I think we can try immediately using the flag `ALLOW_UNASSIGNED`. It would not change anything. except that we collect two exceptions.

**Review by sebastian-nagel — APPROVED** [review]
Thanks, @lfoppiano! Looks good.

Could simplify handling the exceptions, see inline comment.

Merge conflicts need to be resolved.

File: src/main/java/org/commoncrawl/net/HostName.java (hunk 8 above)
Comment (lfoppiano) [reply src/main/java/org/commoncrawl/net/HostName.java]: Thanks. I've changed it accordingly.
